### PR TITLE
Add Group Normalization

### PR DESCRIPTION
**Pull request by lyakaap**
It is an implementation of GroupNormalization.
https://arxiv.org/pdf/1803.08494.pdf 

GN can outperform in tasks such as time-series, GANs, and object detection.
This implementation uses batch normalization internally because we can exploit their optimized implementation with less effort.

### CODE REVIEW

**Comment by crcrpar** [comment]
I wonder why this implementation assumes inputs' dimension is greater than 2 because group normalization will work if the input is 2D.
Though the original paper assumes group normalization is applied to images, some implementations e.g. pytorch accepts 2D input.

Ignore this if wrong and/or ignorable.

**Comment by crcrpar** [comment]
I appreciate your response.

> Or you mean, GN should also accept 2D[batch size, channel size] input?

No, I was just wondering the reason and understood well now.
Thank you.

**Comment by lyakaap** [comment]
No problem. And thanks for pointing.
I'm not sure whether it should accept 1d input or not, I changed error message clearer.

**Comment by tkerola** [comment]
Hi, I saw that `groups` was changed to `n_groups` in this PR, and would like to comment on consistency in the Chainer API. In `Convolution2DFunction`, `groups` is already used to indicate the number of groups used in grouped convolution. As the meaning of groups in this PR is related (each group is processed separately), I would like to hear you opinion on whether both functions should have the same keyword naming in order to make the API consistent? https://github.com/chainer/chainer/blob/master/chainer/functions/connection/convolution_2d.py#L39
If so, perhaps the we should revert to use `groups` instead of `n_groups` in this PR. Thanks.

**Comment by lyakaap** [comment]
I didn't know that and I also think it would be good to fit the naming consistency.
Also in tensorflow, `groups` is used in GN arg name. https://github.com/tensorflow/tensorflow/blob/d0f5bc17560fc97bcc7de9164aa3b237a8d5221d/tensorflow/contrib/layers/python/layers/normalization.py#L167
I will update.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
(Ignore AppVeyor failure)

I observe two failures in our internal test. Could you check them?
One is in multi-gpu test:
```
E   AssertionError: Parameterized test failed.
E   
E   Base test method: GroupNormalizationTest.test_forward_multi_gpu
E   Test parameters:
E     dtype: <type 'numpy.float32'>
E     shape: (5, 4, 15)
E     groups: 1
E   
E   AssertionError: 
E   Fail: 3, Success: 0
E   
E   The first error message:
E   Traceback (most recent call last):
E     File "chainer/testing/condition.py", line 60, in <lambda>
E       lambda: f(ins, *args[1:], **kwargs),
E     File "/work/chainer/tests/chainer_tests/links_tests/normalization_tests/test_group_normalization.py", line 66, in test_forward_multi_gpu
E       self.check_forward(x)
E     File "/work/chainer/tests/chainer_tests/links_tests/normalization_tests/test_group_normalization.py", line 32, in check_forward
E       y = self.link(x_data)
E     File "chainer/links/normalization/group_normalization.py", line 89, in __call__
E       x, self.groups, self.gamma, self.beta, self.eps)
E     File "chainer/functions/normalization/group_normalization.py", line 67, in group_normalization
E       x, dummy_gamma, dummy_beta, eps=eps)
E     File "chainer/functions/normalization/batch_normalization.py", line 798, in batch_normalization
E       axis).apply((x, gamma, beta))[0]
E     File "chainer/function_node.py", line 258, in apply
E       outputs = self.forward(in_data)
E     File "chainer/functions/normalization/batch_normalization.py", line 250, in forward
E       self.inv_std[expander], gamma, beta)
E     File "chainer/functions/normalization/batch_normalization.py", line 694, in _apply_bn_fwd
E       )(x, mean, inv_std, gamma, beta)
E     File "cupy/core/elementwise.pxi", line 580, in cupy.core.core.ElementwiseKernel.__call__
E     File "cupy/core/elementwise.pxi", line 148, in cupy.core.core._preprocess_args
E   ValueError: Array device must be same as the current device: array device = 0 while current = 1
```

Another is documentation. Please add the link to documentation, too.
```
Warning, treated as error:
WARNING: Undocumented entities found.

chainer.links.GroupNormalization
```

**Comment by lyakaap** [comment]
About multi-gpu forward issue:
In multi-gpu test, I followed the test code of batch_normalization,
https://github.com/chainer/chainer/blob/a137f39b2a511ef5ff40a9a363e13a0df651c221/tests/chainer_tests/links_tests/normalization_tests/test_batch_normalization.py#L123
I guess device inconsistency comes from this part, but I would like to know why this part became like this.

**Comment by niboshi** [comment]
Chainer functions and links should perform computation on the device on which input arrays are allocated, no matter what the current device is.

It seems the problem is that `F.group_normalization` does not set the current device to `x.device`.

https://github.com/chainer/chainer/pull/4638/files#diff-9ff199fca61b38b330297137894e651dR61

Here `dummy_gamma` and `dummy_beta` are allocated on the device 0 whereas `x` is on the device 1, causing `F.batch_normalization` to fail.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by lyakaap** [comment]
Thanks for the explanation. I pushed the modification.

File: chainer/functions/normalization/group_normalization.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    if channels % groups != 0:
+        raise ValueError('Argument: \'groups\' must be a divisor '
+                         'of the number of channel.')
+
+    # By doing this reshaping, calling batch_normalization function becomes
+    # equivalent to Group Normalization.
+    # And redundant dimension is added in order to utilize ideep64/cuDNN.
+    x = reshape.reshape(x, (1, batch_size * groups, -1, 1))
+
+    with cuda.get_device_from_array(x):
```
Comment (niboshi) [inline chainer/functions/normalization/group_normalization.py]: `cuda.get_device_from_array` only accept `cupy.ndarray`s.
Please fix it as `with cuda.get_device_from_array(x):`.
(I confirmed the test to pass with this change).

**Comment by niboshi** [comment]
Thanks for a quick fix!
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
Jenkins, test this please.

**Comment by niboshi** [comment]
LGTM!
Thank you very much!

**Pull request merged**